# Advent Of Code 2018 - [Day 1](https://adventofcode.com/2018/day/1)

# Day 1 / Part 1

### Solution Using Pandas library

In [1]:
import pandas as pd

In [2]:
%%timeit
pd.read_csv('data/Day1_Input.txt', header=None, names=['Frequency']).sum()

1.66 ms ± 73.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


### Solution using "pain, vanilla" Python

Some people may not realize it, but there is a built-in ```sum()``` function. I then used list comprehension syntax to parse the values in the input file text as a Python ```list```

In [3]:
%%timeit
sum([int(item.replace('+','')) for item in open('data/Day1_Input.txt')])

411 µs ± 7.31 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# Day 1 / Part 2

Since the puzzle instructions says that I may have to loop through my input data multiple times until I reach my solution, I looked into using itertools' ```cycle``` class

In [4]:
import itertools

freq_list = itertools.cycle([int(item.replace('+','')) for item in open('data/Day1_Input.txt')])

sum = 0
already_seen = set()
for i, value in enumerate(freq_list):
    sum = sum + value
    if sum in already_seen:
        print(i, "ith sum:", sum)
        break
    already_seen.add(sum)
    # Since the for loop would loop forever with a cycled list, we have to stop somewhere
    if i > 1000000:
        print("After a million iterations, I still have not found the answer")
        break

137336 ith sum: 72330


Wow, it took 137,336 iterations to get to my solution. I did not expect it to take this many iterations since my input data only has 960 values.